holiday-management/src/business-logic/holiday-service.ts

```diff
@@ -49,14 +55,56 @@ export class HolidayService {
         await dbService.updateHolidayContainer(container);
 
         holidayRequest.isActive = true;
+        holidayRequest.status = HolidayRequestStatuses.StatusesIds.waitingForApprove;
         return dbService.createHolidayRequest(holidayRequest);
     }
 
+    async cancelHolidayRequest(holidayRequest: HolidayRequest): Promise<HolidayRequest> {
+        let container = await dbService.getHolidayContainer(holidayRequest.userId);
+        let category = container.categories.find((c: HolidayContainerRow) => c.category === holidayRequest.type);
+
+        category.taken -= holidayRequest.days;
+        category.sum += holidayRequest.days;
+
+        await dbService.updateHolidayContainer(container);
+        return holidayRequest;
+    }
+
     getHolidayRequests(userId?: number): Promise<Array<HolidayRequest>> {
         return dbService.getHolidayRequests(userId);
     }
 
     async updateHolidayRequest(holidayRequest: HolidayRequest): Promise<HolidayRequest> {
         return dbService.updateHolidayRequest(holidayRequest);
     }
+
+    async updateHolidayRequestStatus(id: string, status: string): Promise<HolidayRequest> {
+        return new Promise<HolidayRequest>((resolve, reject) => {
+            dbService.getHolidayRequest(id)
+                .then(request => {
+                    let statusObject = HolidayRequestStatuses.getStatusById(status);
+                    if (!statusObject) {
+                        throw `There is no status with name '${status}'`;
+                    }
+
+                    if (appConfig.holidayRequestTransitions[request.status].findIndex(s => s === status) < 0) {
+                        throw `The transition to status '${statusObject.Name}' is prohibited for this holiday request`;
+                    }
+
+                    request.status = status;
+                    return request;
+                })
+                .then(request => dbService.updateHolidayRequest(request))
+                .then(request => {
+                    if (request.status === HolidayRequestStatuses.StatusesIds.cancelledByUser
+                        || request.status === HolidayRequestStatuses.StatusesIds.rejected) {
+                            return this.cancelHolidayRequest(request);
```

Comment:
I wonder if we can make this operation in one transaction with updating status of HolidayRequest. Because right it is possible to introduce inconsistency if this operation fails to whatever reason. To be honest i dont know if transactions are somehow supported in mongo :)

Reply:
Indeed, there might be a problem with integrity of data. In MongoDB there are transactions, but it's possible to implement them:
https://docs.mongodb.com/manual/tutorial/perform-two-phase-commits/
It's a it advanced magic and I'd like to not implement it not, rather later if we will have issues.
Have created "enhancement" to not forget about it:
https://github.com/Wodooman/holiday-management-system/issues/6